usaspending_api/common/logging.py

```diff
@@ -0,0 +1,112 @@
+from django.http import Http404
+from django import http
+from django.utils.timezone import now
+from django.utils.deprecation import MiddlewareMixin
+
+import logging
+import traceback
+
+
+def get_remote_addr(request):
+    ip_address = request.META.get("HTTP_X_FORWARDED_FOR", None)
+    if ip_address:
+        # X_FORWARDED_FOR returns client1, proxy1, proxy2,...
+        ip_address = [x.strip() for x in ip_address.split(",")][0]
+    else:
+        ip_address = request.META.get("REMOTE_ADDR", "")
+
+    return ip_address
+
+
+class LoggingMiddleware(MiddlewareMixin):
+    server_logger = logging.getLogger("server")
+
+    start = None
+    log = None
+
+    def process_request(self, request):
+        """Func called when a request is called on server, function stores request fields for logging"""
+        self.start = now()
+
+        self.log = {
+            "path": request.path,
+            "remote_addr": get_remote_addr(request),
+            "host": request.get_host(),
+            "method": request.method
+        }
+
+        try:
+            # When request body is returned  as <class 'bytes'>
+            self.log["request"] = getattr(request, '_body', request.body).decode('ASCII')
+        except UnicodeDecodeError:
+            self.log["request"] = getattr(request, '_body', request.body)
+
+    def process_response(self, request, response):
+        """
+        Func gets called before reponse is returned from server, stores response fields and logs
+        request and response to server.log
+        """
+
+        status_code = response.status_code
+
+        self.log["status_code"] = status_code
+        self.log["response_ms"] = self.get_response_ms()
+        self.log["traceback"] = None
+
+        if 100 <= status_code < 400:
+            self.log["status"] = 'INFO'
+            self.log["timestamp"] = now().strftime('%d/%m/%y %H:%M:%S')
```

Comment:
Should the timestamp value be moved out of this if statement? 
404 method can overwrite it if necessary

Reply:
Timestamp is required field for all calls. 404 status calls are not included in this condition since 404 status calls are logged as a warning while 100 -399 are logged as INFO

Reply:
Does it make more sense to log the timestamp on request instead of response?